Firestore/core/test/firebase/firestore/testutil/xcgmock.h

```diff
@@ -0,0 +1,210 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_TEST_FIREBASE_FIRESTORE_TESTUTIL_XCGMOCK_H_
+#define FIRESTORE_CORE_TEST_FIREBASE_FIRESTORE_TESTUTIL_XCGMOCK_H_
+
+#import <XCTest/XCTest.h>
+
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <utility>
+
+#import "Firestore/Source/Model/FSTDocument.h"
+
+#include "Firestore/core/src/firebase/firestore/util/string_apple.h"
+#include "gmock/gmock.h"
+
+namespace firebase {
+namespace firestore {
+namespace testutil {
+
+template <typename M>
+class XcTestRecorder {
+ public:
+  XcTestRecorder(M matcher, XCTestCase* test_case, const char* file, int line)
+      : formatter_(std::move(matcher)),
+        test_case_(test_case),
+        file_(file),
+        line_(line) {
+  }
+
+  template <typename T>
+  void Match(const char* text, const T& value) const {
+    testing::AssertionResult result = formatter_(text, value);
+    if (!result) {
+      RecordFailure(result.message());
+    }
+  }
+
+  void RecordFailure(const char* message) const {
+    [test_case_
+        recordFailureWithDescription:[NSString stringWithUTF8String:message]
+                              inFile:[NSString stringWithUTF8String:file_]
+                              atLine:line_
+                            expected:YES];
+  }
+
+ private:
+  testing::internal::PredicateFormatterFromMatcher<M> formatter_;
+
+  XCTestCase* test_case_;
+  const char* file_;
+  int line_;
+};
+
+template <typename M>
+XcTestRecorder<M> MakeXcTestRecorder(M matcher,
+                                     XCTestCase* test_case,
+                                     const char* file,
+                                     int line) {
+  return XcTestRecorder<M>(std::move(matcher), test_case, file, line);
+}
+
+#define XC_ASSERT_THAT(actual, matcher)                                \
+  do {                                                                 \
+    using actual_type =                                                \
+        typename std::remove_reference<decltype(actual)>::type;        \
+    auto recorder = firebase::firestore::testutil::MakeXcTestRecorder( \
+        matcher, self, __FILE__, __LINE__);                            \
+    recorder.Match(#actual, actual);                                   \
+  } while (0)
+
+/**
+ * Prints the -description of an Objective-C object to the given ostream.
+ */
+inline void ObjcPrintTo(id value, std::ostream* os) {
+  // Force the result type to NSString* or we can't resolve MakeString.
+  NSString* description = [value description];
+  *os << util::MakeString(description);
```

Comment:
I'm not 100% sure `util::ToString` (from `to_string.h`) works with `id` parameters, but if it does, perhaps use it here?
```cpp
*os << util::ToString(value);
```